src/core/processing/qgsprocessingutils.cpp

```diff
@@ -172,3 +173,29 @@ QString QgsProcessingUtils::normalizeLayerSource( const QString &source )
   return normalized.trimmed();
 }
 
+QgsFeatureIterator QgsProcessingUtils::getFeatures( QgsVectorLayer *layer, const QgsProcessingContext &context, const QgsFeatureRequest &request )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
+
+  QgsFeatureRequest req( request );
+  req.setInvalidGeometryCheck( context.invalidGeometryCheck() );
+  req.setInvalidGeometryCallback( context.invalidGeometryCallback() );
+  if ( useSelection )
+  {
+    return layer->selectedFeaturesIterator( req );
+  }
+  else
+  {
+    return layer->getFeatures( req );
+  }
+}
+
+long QgsProcessingUtils::featureCount( QgsVectorLayer *layer, const QgsProcessingContext &context )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
```

Comment:
I would appreciate getting rid of the current behavior and replace it with a superior one :) -- I find that a processing-wide single [x] use only selected flag is really bad. Beyond offering absolutely no visible feedback, it's not practical. 

I'd rather have some sort of a checkbox attached to the layer selection widget, which would offer a use-selected-only UI at hand.